toucan_connectors/hubspot_private_app/hubspot_connector.py

```diff
@@ -0,0 +1,150 @@
+from datetime import datetime
+from enum import Enum
+from typing import Any, Generator, Protocol, TypeAlias
+
+import pandas as pd
+from hubspot import HubSpot  # type:ignore[import]
+from pydantic import BaseModel, Field, SecretStr
+
+from toucan_connectors.pagination import build_pagination_info
+from toucan_connectors.toucan_connector import DataSlice, ToucanConnector, ToucanDataSource
+
+
+class HubspotDataset(str, Enum):
+    contacts = 'contacts'
+    companies = 'companies'
+    deals = 'deals'
+    quotes = 'quotes'
+    owners = 'owners'
+
+
+class HubspotDataSource(ToucanDataSource):
+    dataset: HubspotDataset
+
+
+class _HubSpotPagingNext(BaseModel):
+    after: str
+    link: str
+
+
+class _HubSpotPaging(BaseModel):
+    next_: _HubSpotPagingNext = Field(..., alias='next')
+
+
+class _HubSpotResult(BaseModel):
+    created_at: datetime | None
+    updated_at: datetime | None
+    id_: str = Field(..., alias='id')
+    properties: dict[str, Any] = Field(default_factory=dict)
+
+    def to_dict(self) -> dict[str, Any]:
+        return {
+            'created_at': self.created_at,
+            'updated_at': self.updated_at,
+            'id': self.id_,
+            **self.properties,
+        }
+
+
+class _HubSpotResponse(BaseModel):
+    paging: _HubSpotPaging | None = None
+    results: list[_HubSpotResult]
+
+    def next_page_after(self) -> str | None:
+        return None if self.paging is None else self.paging.next_.after
+
+
+_RawHubSpotResult: TypeAlias = dict[str, str | None | list[Any]]
+
+
+class _HubSpotObject(Protocol):  # pragma: no cover
+    def to_dict(self) -> _RawHubSpotResult:
+        ...
+
+
+class _BasicApi(Protocol):  # pragma: no cover
+    def get_page(self, after: str | None, limit: int | None) -> _HubSpotObject:
+        ...
+
+
+class _Api(Protocol):  # pragma: no cover
+    @property
+    def basic_api(self) -> _BasicApi:
+        ...
+
+    def get_all(self) -> list[_HubSpotObject]:
+        ...
```

Comment:
If i understand correctly this are the object from the hubspot api ?

Would this make sense to move this in another file and create utility function to manipulate them ?

Reply:
There are two things here:
* The pydantic objects are indeed used to parse the hubspot API response
* The `Protocol` objects are pure typing: Since the hubspot client does not provide type stubs, these are type definitions for it 

They _could_ be moved to a submodule, but since this file is so short, I figured that it would be more comfortable for reading to have everything in the same place. No strong opinion on this though